image-loader/app/model/ImageUpload.scala

```diff
@@ -28,57 +29,92 @@ case object ImageUpload {
-        s3Thumb          <- thumbStoreFuture
-        sourceDimensions <- sourceDimensionsFuture
-        thumbDimensions  <- thumbDimensionsFuture
-        fileMetadata     <- fileMetadataFuture
-        colourModel      <- colourModelFuture
-        fullFileMetadata  = fileMetadata.copy(colourModel = colourModel)
-
-        metadata      = ImageMetadataConverter.fromFileMetadata(fullFileMetadata)
-        cleanMetadata = metadataCleaners.clean(metadata)
-
-        sourceAsset = Asset.fromS3Object(s3Source, sourceDimensions)
-        thumbAsset  = Asset.fromS3Object(s3Thumb,  thumbDimensions)
-
-        baseImage      = createImage(uploadRequest, sourceAsset, thumbAsset, fullFileMetadata, cleanMetadata)
-        processedImage = SupplierProcessors.process(baseImage)
-
-        // FIXME: dirty hack to sync the originalUsageRights and originalMetadata as well
-        finalImage     = processedImage.copy(
-          originalMetadata    = processedImage.metadata,
-          originalUsageRights = processedImage.usageRights
-        )
-      }
-      yield {
-        ImageUpload(uploadRequest, finalImage)
+    for {
+      fileMetadata <- fileMetadataFuture
+    } yield {
+
+      val uploadIsPng24 = isPng24(uploadRequest.mimeType, fileMetadata)
+
+      // These futures are started outside the for-comprehension, otherwise they will not run in parallel
+      val sourceStoreFuture = storeSource(uploadRequest)
+      // FIXME: pass mimeType
+      val colourModelFuture = ImageOperations.identifyColourModel(uploadedFile, "image/jpeg")
+      val sourceDimensionsFuture = FileMetadataReader.dimensions(uploadedFile, uploadRequest.mimeType)
+
+      val thumbFuture = for {
+        fileMetadata <- fileMetadataFuture
+        colourModel <- colourModelFuture
+        iccColourSpace = FileMetadataHelper.normalisedIccColourSpace(fileMetadata)
+        thumb <- ImageOperations.createThumbnail(uploadedFile, thumbWidth, thumbQuality, tempDir, iccColourSpace, colourModel)
+      } yield thumb
+
+      val pngOption = if (uploadIsPng24) {
+        val optimisedPng = tempDir.getAbsolutePath() + "/optimisedpng" + ".png"
+        Seq("pngquant", "--quality", "1-85", uploadedFile.getAbsolutePath(), "--output", optimisedPng).!
+
+        Some(new File(optimisedPng))
+      } else
+        None
+
+      bracket(thumbFuture)(_.delete) { thumb =>
+        // Run the operations in parallel
+        val thumbStoreFuture = storeThumbnail(uploadRequest, thumb)
+
+        val pngStoreFuture = if (uploadIsPng24)
+            Some(storeOptimisedPng(uploadRequest, pngOption.get)).map(result => result.map(Option(_)))
+              .getOrElse(Future.successful(None))
+            else
+              Future(None)
+
+        val thumbDimensionsFuture = FileMetadataReader.dimensions(thumb, Some("image/jpeg"))
+
+        val pngDimensionsFuture = if (uploadIsPng24)
+            FileMetadataReader.dimensions(pngOption.get, Some("image/png"))
+          else
+            Future(None)
```

Comment:
Need another Scala person's feedback, but i think this can just be "None", it will get overlooked when used in the "for" comprehension below. 

@akash1810 is that about right? This is for my own understanding, not a blocker